compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -1131,8 +1131,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let mut err = self.demand_suptype_diag(expr.span, expected_ty, actual_ty).unwrap();
             let lhs_ty = self.check_expr(lhs);
             let rhs_ty = self.check_expr(rhs);
+            let might_coerce = |lhs: Ty<'tcx>, rhs: Ty<'tcx>| {
```

Comment:
This helper is named pretty vaguely -- it's not checking if lhs and rhs might coerce, it's checking if the lhs and rhs's refs can coerce.

```suggestion
            let refs_can_coerce = |lhs: Ty<'tcx>, rhs: Ty<'tcx>| {
```

Reply:
Yeah. It's weird because from the user's point of view it is `Ty`, but auto-deref is at play in binops (which is why I peel *all* refs and add one layer after).